### PR TITLE
issue 526.1: initial support for vocbase-prefixed analyzer names

### PR DESCRIPTION
**Pull request by vasiliy-arangodb**

### CODE REVIEW

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 1]
```diff
@@ -57,6 +58,7 @@
 namespace {
 
 static std::string const ANALYZER_COLLECTION_NAME("_iresearch_analyzers");
+static char const ANALYZER_PREFIX_DELIM = ':'; // name prefix delimiter (2 chars)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: the comment says `2 chars`, put the delimiter is just one char long...

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 2]
```diff
@@ -1085,6 +1142,53 @@ bool IResearchAnalyzerFeature::loadConfiguration() {
...
+  irs::string_ref const& name, // analyzer name
+  TRI_vocbase_t const& activeVocbase, // fallback vocbase if not part of name
+  TRI_vocbase_t const& systemVocbase, // the system vocbase for use with empty prefix
+  bool expandVocbasePrefix /*= true*/ // use full vocbase name as prefix for active/system v.s. EMPTY/'::'
+) {
+  if (IDENTITY_ANALYZER_NAME == name) {
+    return name; // special case for the 'identity' analyzer which is a singleton
+  }
+
+  // FIXME TODO remove once lagacy analyzers will be added to each vocbase via an upgrade step and JavaScript tests updated accordingly
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]:
```suggestion
  // FIXME TODO remove once legacy analyzers will be added to each vocbase via an upgrade step and JavaScript tests updated accordingly
```

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 3]
```diff
@@ -1085,6 +1142,53 @@ bool IResearchAnalyzerFeature::loadConfiguration() {
...
+  {
+    static std::unordered_set<std::string> legacyAnalyzers = {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: can we make this set const?
```suggestion
    static std::unordered_set<std::string> const legacyAnalyzers = {
```

File: arangod/IResearch/IResearchLinkHelper.cpp [hunk 4]
```diff
@@ -200,14 +198,12 @@ arangodb::Result modifyLinks(std::unordered_set<TRI_voc_cid_t>& modified,
...
+  for (auto& id: stale) {
     if (!trxCtx->resolver().getCollection(id)) {
       LOG_TOPIC(WARN, arangodb::iresearch::TOPIC)
-          << "request for removal of a stale link to a missing collection '"
-          << id << "', ignoring";
+        << "request for removal of a stale link to a missing collection '" << id << "', ignoring";
 
-      continue;  // skip adding removal requests to stale links to non-existant
-                 // collections (already dropped)
+      continue; // skip adding removal requests to stale links to non-existant collections (already dropped)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLinkHelper.cpp]:
```suggestion
      continue; // skip adding removal requests to stale links to non-existent collections (already dropped)
```

File: arangod/IResearch/IResearchLinkMeta.cpp [hunk 5]
```diff
@@ -172,10 +173,13 @@ bool IResearchLinkMeta::operator!=(IResearchLinkMeta const& other) const noexcep
...
-                             IResearchLinkMeta const& defaults /*= DEFAULT()*/,
-                             Mask* mask /*= nullptr*/
-                             ) noexcept {
+bool IResearchLinkMeta::init( // initialize meta
+    arangodb::velocypack::Slice const& slice, // definition
+    std::string& errorField, // field causing error (out-param)
+    IResearchLinkMeta const& defaults /*= DEFAULT()*/, // inherited defaults
+    TRI_vocbase_t const* defaultVocbase /*= nullptr*/, // fallback vocbase
+    Mask* mask /*= nullptr*/ // initialized fields (out-param)
+) noexcept {
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLinkMeta.cpp]: Technically, this method can throw exceptions, e.g. when initially constructing the static string object `fieldName`, when filling `errorField` etc.
So I suggest not declaring it `noexcept`.

File: arangod/IResearch/IResearchLinkMeta.cpp [hunk 6]
```diff
@@ -381,8 +404,27 @@ bool IResearchLinkMeta::json(arangodb::velocypack::Builder& builder,
...
     for (auto& entry : _analyzers) {
-      if (entry) {  // skip null analyzers
-        analyzersBuilder.add(arangodb::velocypack::Value(entry->name()));
+      if (!entry) {
+        continue; // skip null analyzers
+      }
+
+      if (defaultVocbase) {
+        auto* sysDatabase = arangodb::application_features::ApplicationServer::lookupFeature< // find feature
+          arangodb::SystemDatabaseFeature // featue type
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLinkMeta.cpp]:
```suggestion
          arangodb::SystemDatabaseFeature // feature type
```

File: arangod/IResearch/IResearchLinkMeta.cpp [hunk 7]
```diff
@@ -391,30 +433,28 @@ bool IResearchLinkMeta::json(arangodb::velocypack::Builder& builder,
...
-
-      subDefaults._fields.clear();  // do not inherit fields and overrides
-                                    // overrides from this field
+    subDefaults._fields.clear(); // do not inherit fields and overrides from this field
+    fieldsBuilder.openObject();
 
       for (auto& entry : _fields) {
-        mask._fields = !entry.value()->_fields.empty();  // do not output empty fields on subobjects
+        fieldMask._fields = !entry.value()->_fields.empty(); // do not output empty fields on subobjects
+        fieldBuilder.openObject();
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLinkMeta.cpp]: IMHO using a second Builder instance would not be necessary here.
The entire data could also be built in a single Builder instance, even if it is nested.

File: arangod/V8Server/v8-views.cpp [hunk 8]
```diff
@@ -42,16 +42,12 @@
 namespace {
 
 ////////////////////////////////////////////////////////////////////////////////
-/// @return the specified object is granted 'level' access
+/// @return the specified vocbase is granted 'level' access
 ////////////////////////////////////////////////////////////////////////////////
-bool canUse(arangodb::auth::Level level, TRI_vocbase_t const& vocbase,
-            std::string const* dataSource = nullptr  // nullptr == validate only vocbase
-) {
+bool canUse(arangodb::auth::Level level, TRI_vocbase_t const& vocbase) {
```
Comment (jsteemann) [inline arangod/V8Server/v8-views.cpp]: is this still needed?

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 1 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: true, it's simpler to compare 1 char at a time

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 2 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: it'll be removed in the next pull-request

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 3 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: it'll be removed in the next pull-request

File: arangod/V8Server/v8-views.cpp (hunk 8 above)
Comment (vasiliy-arangodb) [reply arangod/V8Server/v8-views.cpp]: yes, this is for LogicalView creation where there is not LogicalView object yet, I haven't figured out a clean way to remove it

File: arangod/IResearch/IResearchLinkHelper.cpp (hunk 4 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchLinkHelper.cpp]: done

File: arangod/IResearch/IResearchLinkMeta.cpp (hunk 5 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchLinkMeta.cpp]: done

File: arangod/IResearch/IResearchLinkMeta.cpp (hunk 6 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchLinkMeta.cpp]: done

File: arangod/IResearch/IResearchLinkMeta.cpp (hunk 7 above)
Comment (vasiliy-arangodb) [reply arangod/IResearch/IResearchLinkMeta.cpp]: done

**Comment by vasiliy-arangodb** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3321/

**Pull request merged**